File: mgyoutube-multitech/api-java/src/main/java/com/djpedesen/mgyoutube/api_java/webservices/SupportWebService.java
package com.djpedesen.mgyoutube.api_java.webservices;

import java.net.URISyntaxException;

import javax.ws.rs.Consumes;
import javax.ws.rs.DELETE;
import javax.ws.rs.GET;
import javax.ws.rs.PUT;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import javax.ws.rs.core.Response.Status;

import com.djpedesen.mgyoutube.api_java.apimodel.User;
import com.djpedesen.mgyoutube.api_java.modules.UserModule;
import com.google.gson.Gson;

@Path("/support")
public class SupportWebService {
	private static final Gson GSON = new Gson();

	private UserModule userModule;

	public SupportWebService() {
		this.userModule = ModuleRepoRegistry.getUserModule();
	}

	@GET
	@Path("/users/{username}")
	@Produces(MediaType.APPLICATION_JSON)
	public Response getUserByUsername(@PathParam(value = "username") final String username) {
		System.out.println("getUserByUsername: username=" + username);

		// TODO: need to sanitize payload input before using
		final String santizedUsername = username;
		final User user = userModule.getUser(santizedUsername);
		System.out.println("getUserByUsername: user=" + user);

		if (user == null) {
			System.out.println("getUserByUsername: username=" + santizedUsername + " returning NOT_FOUND");
			return Response.status(Status.NOT_FOUND).build();
		}

		user.password = null;

		final String responseJson = GSON.toJson(user);

		System.out.println("getUserByUsername: username=" + santizedUsername + " returning OK");
		return Response.ok(responseJson, MediaType.APPLICATION_JSON).build();
	}

	@PUT
	@Path("/users/{username}")
	@Consumes(MediaType.APPLICATION_JSON)
	@Produces(MediaType.APPLICATION_JSON)
	public Response createUpdateUserByUsername(@PathParam(value = "username") final String username,
			final String userJson) throws URISyntaxException {
		System.out.println("createUpdateUserByUsername: username=" + username + " userJson=" + userJson);

		// TODO: need to sanitize payload input before using
		// final String sanitizedUsername = username;
		final String sanitizedUserJson = userJson;
		final User user = Helpers.marshalUserFromJson(sanitizedUserJson);
		System.out.println("createUpdateUserByUsername: user=" + user);

		final User createdUser = userModule.createUpdateUser(user);
		createdUser.password = null;
		final String createdUserJson = GSON.toJson(createdUser);

		// final URI location = new URI("/users/" + createdUser.userId);

		System.out.println(
				"createUpdateUserByUsername: userJson=" + sanitizedUserJson + " returning OK with " + createdUserJson);
		return Response.ok(createdUserJson, MediaType.APPLICATION_JSON).build();
	}

	@DELETE
	@Path("/users/{username}")
	public Response deleteUserByUsername(@PathParam(value = "username") final String username) {
		System.out.println("deleteUserByUsername: username=" + username);

		// TODO: need to sanitize payload input before using
		final String santizedUsername = username;
		final User oldUser = userModule.removeUser(santizedUsername);

		if (oldUser == null) {
			System.out.println("deleteUserByUsername: username=" + santizedUsername + " returning NOT_FOUND");
			return Response.status(Status.NOT_FOUND).build();
		}

		return Response.ok().build();
	}

}

File: mgyoutube-multitech/api-dotnet/webservices/SupportWebService.cs
using System;
using System.IO;
using System.Threading.Tasks;
using api_dotnet.apimodel;
using api_dotnet.aspnet;
using api_dotnet.modules;
using Microsoft.AspNetCore.Http;
using Microsoft.AspNetCore.Routing;
using Newtonsoft.Json;

namespace api_dotnet.webservices
{
    public class SupportWebService
    {
        private UserModule userModule;

        public SupportWebService()
        {
            Console.WriteLine("Creating SupportWebService");
            this.userModule = ModuleRepoRegistry.TheUserModule;
        }

        public void SetupRoutes(IRouteBuilder routeBuilder)
        {
            routeBuilder.MapPut("/api/support/users/{username}", CreateUser);
            routeBuilder.MapGet("/api/support/users/{username}", GetUserByUsername);
            routeBuilder.MapDelete("/api/support/users/{username}", DeleteUserByUsername);
        }

        public Task CreateUser(HttpContext context)
        {
            try
            {
                string userJson = RequestHelper.GetRequestBody(context.Request);
                Console.WriteLine("createUser: userJson=" + userJson);

                // TODO: need to sanitize payload input before using
                string sanitizedUserJson = userJson;
                User user = Helpers.MarshalUserFromJson(sanitizedUserJson);
                Console.WriteLine("createUser: user=" + user);

                User createdUser = userModule.CreateUpdateUser(user).GetAwaiter().GetResult();
                createdUser.password = null;

                string responseJson = JsonConvert.SerializeObject(createdUser);

                Console.WriteLine("createUser: userJson=" + userJson + " returning OK and " + responseJson);
                return ResponseHelper.Ok(context.Response, responseJson, MediaType.APPLICATION_JSON);
            }
            catch (Exception e)
            {
                Console.WriteLine("CreateUser, unhandled exception", e);
                return ResponseHelper.InternalServerError(context);
            }
        }


        public Task GetUserByUsername(HttpContext context)
        {
            try
            {
                string username = (string)context.GetRouteValue("username");
                Console.WriteLine("getUserByUsername: username=" + username);

                // TODO: need to sanitize payload input before using
                string santizedUsername = username;
                User user = userModule.GetUser(santizedUsername).GetAwaiter().GetResult();
                Console.WriteLine("getUserByUsername: user=" + user);

                if (user == null)
                {
                    Console.WriteLine("getUserByUsername: username=" + username + " returning NOT_FOUND");
                    return ResponseHelper.NotFound(context);
                }

                user.password = null;

                string responseJson = JsonConvert.SerializeObject(user);

                Console.WriteLine("getUserByUsername: username=" + username + " returning OK with " + responseJson);
                return ResponseHelper.Ok(context.Response, responseJson, MediaType.APPLICATION_JSON);
            }
            catch (Exception e)
            {
                Console.WriteLine("GetUserByUsername, unhandled exception", e);
                return ResponseHelper.InternalServerError(context);
            }
        }


        public Task DeleteUserByUsername(HttpContext context)
        {
            try
            {
                string username = (string)context.GetRouteValue("username");
                Console.WriteLine("deleteUserByUsername: username=" + username);

                // TODO: need to sanitize payload input before using
                string santizedUsername = username;
                User oldUser = userModule.RemoveUser(santizedUsername).GetAwaiter().GetResult();

                if (oldUser == null)
                {
                    Console.WriteLine("deleteUserByUsername: username=" + username + " returning NOT_FOUND");
                    return ResponseHelper.NotFound(context);
                }

                return ResponseHelper.Ok(context);
            }
            catch (Exception e)
            {
                Console.WriteLine("DeleteUserByUsername, unhandled exception", e);
                return ResponseHelper.InternalServerError(context);
            }
        }
    }
}
